DOCS-4086: code samples for building a good dataset (#4413)
viamrobotics/docs: docs/data-ai/train/capture-images.md

Reviewer: Shouldn't this page mention some of the stuff on/link to Upload external data for training, such as adding images from mobile? I personally have found that method much easier for certain cases

Reviewer 2: I was thinking that too Jessamy but yeah doing that separately seems good

---
linkTitle: "Capture images for training"
title: "Capture images for training"
weight: 20
layout: "docs"
type: "docs"
description: "Capture images that you can use to train a machine learning model."
---

## Prerequisites

{{% expand "A machine connected to Viam" %}}

{{% snippet "setup.md" %}}

{{% /expand %}}

{{% expand "A camera, connected to your machine, to capture images" %}}

Follow the guide to configure a [webcam](/operate/reference/components/camera/webcam/) or similar [camera component](/operate/reference/components/camera/).

{{% /expand%}}

{{< alert title="Tip" color="tip" >}}

For the best results, use the same camera for both training data capture and production deployment.

{{< /alert >}}

## Capture individual images

{{< tabs >}}
{{% tab name="Web UI" %}}

You can add images to a dataset directly from a camera or vision component feed in the machine's **CONTROL** or **CONFIGURATION** tabs.

To add an image directly to a dataset from a visual feed, complete the following steps:

> 1. Open the **TEST** panel of any camera or vision service component to view a feed of images from the camera.

Reviewer: Seemed not to work for me if I had it on live--maybe a bug/button should be disabled if live. Are you already on this? Otherwise LMK if I should make an eng ticket
```suggestion
1. Open the **TEST** panel of any camera or vision service component to view a feed of images from the camera.
1. Use the refresh frequency dropdown to select an option _other than_ **Live**.
```

Author: Works for me, so if you're seeing an issue, you should file a ticket. (un?)fortunately it all works for me properly.

Reviewer: Okay; I messaged team-viz (since this was persistent the other day but maybe they already fixed it since I can't repro)

{{% /tab %}}
{{% tab name="Python" %}}

To capture an image and add it to your **DATA** page, fetch an image from your camera through your machine.
Pass that image and an appropriate set of metadata to [`data_client.binary_data_capture_upload`](/dev/reference/apis/data-client/#binarydatacaptureupload):

```python
CAMERA_NAME = "<camera-name>"
MACHINE_ADDRESS = "<machine-address.viam.cloud>"

dial_options = DialOptions(
    credentials=Credentials(
        type="api-key",
        payload=API_KEY,
    ),
    auth_entity=API_KEY_ID,
)

robot_opts = RobotClient.Options.with_api_key(
    api_key=API_KEY,
    api_key_id=API_KEY_ID
)

viam_client = await ViamClient.create_from_dial_options(dial_options)
data_client = viam_client.data_client

robot_client = await RobotClient.at_address(ROBOT_ADDRESS, robot_opts)
camera = Camera.from_robot(robot_client, CAMERA_NAME)

# Capture image
image_frame = await camera.get_image()

# Upload data
file_id = await data_client.binary_data_capture_upload(
    part_id=PART_ID,
    component_type="camera",
    component_name=CAMERA_NAME,

>     method_name="GetImages",

Reviewer: Wait why are all of these saying the method name is GetImages instead of GetImage? Shouldn't it match the method you're using on line 78? Same applies to other language samples.

Author: Good point, I must have forgotten to update this part since it's declared in multiple places. Thanks!

```suggestion
    method_name="GetImage",
```

    BinaryData:        img,
})
```

{{% /tab %}}
{{% tab name="TypeScript" %}}

To capture an image and add it to your **DATA** page, fetch an image from your camera through your machine.
Pass that image and an appropriate set of metadata to [`dataClient.binaryDataCaptureUpload`](/dev/reference/apis/data-client/#binarydatacaptureupload):

```typescript
const CAMERA_NAME = "<camera-name>";
const MACHINE_ADDRESS = "<machine-address.viam.cloud>";
const API_KEY = "<api-key>";
const API_KEY_ID = "<api-key-id>";
const PART_ID = "<part-id>";

const machine = await Viam.createRobotClient({
  host: MACHINE_ADDRESS,
  credential: {
    type: "api-key",
    payload: API_KEY,
  },
  authEntity: API_KEY_ID,
});

const client: ViamClient = await createViamClient({
  credential: {
    type: "api-key",
    payload: API_KEY,
  },
  authEntity: API_KEY_ID,
});

const dataClient = client.dataClient;

const camera = new Viam.CameraClient(machine, CAMERA_NAME);

// Capture image
const imageFrame = await camera.getImage();

// Upload binary data
const now = new Date();
const fileId = await dataClient.binaryDataCaptureUpload({
  partId: PART_ID,
  componentType: "camera",
  componentName: CAMERA_NAME,

>   methodName: "GetImages",

Author: ```suggestion
  methodName: "GetImage",
```

    apiKeyId: API_KEY_ID,
    apiKey: API_KEY,
);

final dataClient = client.dataClient;

final camera = Camera.fromRobot(machine, CAMERA_NAME);

// Capture image
final imageFrame = await camera.getImage();

// Upload binary data
final now = DateTime.now().toUtc();
final fileId = await dataClient.binaryDataCaptureUpload(
    partId: PART_ID,
    componentType: 'camera',
    componentName: CAMERA_NAME,
    methodName: 'GetImages',
    dataRequestTimes: [now, now],
    fileExtension: '.jpg',
    binaryData: imageFrame,
);

// Cleanup
await robotClient.close();
dataClient.close();
```

{{% /tab %}}
{{< /tabs >}}

Once you've captured [enough images for training](/data-ai/train/train-tflite/), you must [annotate](/data-ai/train/annotate-images/) the images before you can use them to train a model.

## Capture images over time

To capture a large number of images for training an ML model, use the data management service to [capture and sync image data](/data-ai/capture-data/capture-sync/) from your camera.

When you sync with data management, Viam stores the images saved by capture and sync on the [**DATA** page](https://app.viam.com/data/), but does not add the images to a dataset.

> To use your captured images for training, [annotate the images](/data-ai/train/annotate-images/), then [add the images to a dataset](/data-ai/train/update-dataset/).

Reviewer: ```suggestion
To use your captured images for training, [add the images to a dataset](/data-ai/train/update-dataset/) and [annotate them](/data-ai/train/annotate-images/), so you can use them to train a model.
```
To follow order in left nav (I guess it doesn't really matter which order they do it in? Just trying to make it slightly less confusing)

{{< alert title="Tip" color="tip" >}}

Once you have enough images, consider disabling data capture to [avoid incurring fees](https://www.viam.com/product/pricing) for capturing large amounts of training data.

{{< /alert >}}

> 
> Once you've captured [enough images for training](/data-ai/train/train-tflite/), you must [annotate](/data-ai/train/annotate-images/) the images before you can use them to train a model.

Reviewer: ```suggestion
```
Redundant with above
